File: lab2no2.cpp
#include <iostream>
#include <string>

using namespace std;

void illustration(int banki, int ostatok, int count_steps) { // вывод иллюстрации
    if (count_steps == 0) { // дано
        cout << "Jars" << endl;
        for (int i = 0; i < banki; ++i) {
            cout << 'F';
        }
        cout << endl;
    } else if (count_steps % 2 != 0) { // для нечетных шагов
        cout << count_steps << " step" << endl;
        for (int i = 0; i < banki; ++i) {
            cout << 'E';
        }
        cout << endl;
    } else if (count_steps % 2 == 0) { // для четных шагов
        cout << count_steps << " step" << endl;
        for (int i = 0; i < banki - ostatok; ++i) {
            cout << 'F';
        }
        cout << endl;
        for (int i = 0; i < ostatok; ++i) {
            cout << 'E';
        }
        cout << endl;
    }
}

int main() {
    int banki, obmen; // ввод данных
    cout << "Input jar quantity: ";
    cin >> banki;
    cout << "Input exchange: ";
    cin >> obmen;

    if (banki <= 0 || obmen <= 0) { // проверка на правильность значений
        cout << "Error" << endl;
    } else {
        int count_empty = 0; // счетчик пустых банок, шагов, оставшихся пустых банок после обмена
        int count_steps = 0;
        int ostatok = 0;

        while (banki >= obmen) { // вывод дано -> +1 шаг -> выпить все банки -> добавить выпитые банки к счетки без учета выпитых ранее -> обмен банок -> +1 шаг 
            if (count_steps == 0) {
                illustration(banki, ostatok, count_steps);
            }
            count_steps++;
            illustration(banki, ostatok, count_steps);
            count_empty += banki - ostatok;
            ostatok = banki % obmen;
            banki = (banki / obmen) + ostatok;
            count_steps++;
            illustration(banki, ostatok, count_steps);
        }

        if (banki < obmen) { // последний шаг, когда нельзя обменять банки
            count_empty += banki - ostatok;
            count_steps++;
            illustration(banki, ostatok, count_steps);
            cout << "Jars drunk: " << count_empty << endl;
            cout << "Step count: " << count_steps << endl;
        }
    }

    return 0;
}

File: lab2no2.cs
using System;

class Program
{
    static void Illustration(int banki, int ostatok, int count_steps)
    {
        if (count_steps == 0)
        {
            Console.WriteLine("Jars");
            for (int i = 0; i < banki; i++)
            {
                Console.Write("F");
            }
            Console.WriteLine();
        }
        else if (count_steps % 2 != 0)
        {
            Console.WriteLine($"{count_steps} step");
            for (int i = 0; i < banki; i++)
            {
                Console.Write("E");
            }
            Console.WriteLine();
        }
        else
        {
            Console.WriteLine($"{count_steps} step");
            for (int i = 0; i < (banki - ostatok); i++)
            {
                Console.Write("F");
            }
            Console.WriteLine();
            for (int i = 0; i < ostatok; i++)
            {
                Console.Write("E");
            }
            Console.WriteLine();
        }
    }

    static void Main()
    {
        Console.WriteLine("Input jar quantity: ");
        int banki = int.Parse(Console.ReadLine());

        Console.WriteLine("Input exchange: ");
        int obmen = int.Parse(Console.ReadLine());

        if (banki <= 0 || obmen <= 0)
        {
            Console.WriteLine("Error");
        }
        else
        {
            int count_empty = 0;
            int count_steps = 0;
            int ostatok = 0;
            int new_banki = banki;

            while (new_banki >= obmen)
            {
                if (count_steps == 0)
                {
                    Illustration(new_banki, ostatok, count_steps);
                }
                count_steps++;
                Illustration(new_banki, ostatok, count_steps);
                count_empty += new_banki - ostatok;
                ostatok = new_banki % obmen;
                new_banki = (new_banki / obmen) + ostatok;
                count_steps++;
                Illustration(new_banki, ostatok, count_steps);
            }

            if (new_banki < obmen)
            {
                count_empty += new_banki - ostatok;
                count_steps++;
                Illustration(new_banki, ostatok, count_steps);
                Console.WriteLine($"Jars drunk: {count_empty}");
                Console.WriteLine($"Step count: {count_steps}");
            }
        }
    }
}
